src/thread.rs

```diff
@@ -0,0 +1,583 @@
+const MAIN_THREAD: ThreadId = ThreadId(0);
+
+impl Idx for ThreadId {
+    fn new(idx: usize) -> Self {
+        ThreadId(idx)
+    }
+    fn index(self) -> usize {
+        self.0
+    }
+}
+
+impl From<u64> for ThreadId {
+    fn from(id: u64) -> Self {
+        Self(usize::try_from(id).unwrap())
+    }
+}
+
+impl From<u32> for ThreadId {
+    fn from(id: u32) -> Self {
+        Self(usize::try_from(id).unwrap())
+    }
+}
+
+impl ThreadId {
+    pub fn to_u32_scalar<'tcx>(&self) -> Scalar<Tag> {
+        Scalar::from_u32(u32::try_from(self.0).unwrap())
+    }
+}
+
+/// An identifier of a set of blocked threads.
+#[derive(Clone, Copy, Debug, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct BlockSetId(NonZeroU32);
+
+impl From<u32> for BlockSetId {
+    fn from(id: u32) -> Self {
+        Self(NonZeroU32::new(id).expect("0 is not a valid blockset id"))
+    }
+}
+
+impl BlockSetId {
+    pub fn to_u32_scalar<'tcx>(&self) -> Scalar<Tag> {
+        Scalar::from_u32(self.0.get())
+    }
+}
+
+/// The state of a thread.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum ThreadState {
+    /// The thread is enabled and can be executed.
+    Enabled,
+    /// The thread tried to join the specified thread and is blocked until that
+    /// thread terminates.
+    BlockedOnJoin(ThreadId),
+    /// The thread is blocked and belongs to the given blockset.
+    Blocked(BlockSetId),
+    /// The thread has terminated its execution (we do not delete terminated
+    /// threads).
+    Terminated,
+}
+
+/// The join status of a thread.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum ThreadJoinStatus {
+    /// The thread can be joined.
+    Joinable,
+    /// A thread is detached if its join handle was destroyed and no other
+    /// thread can join it.
+    Detached,
+    /// The thread was already joined by some thread and cannot be joined again.
+    Joined,
+}
+
+/// A thread.
+pub struct Thread<'mir, 'tcx> {
+    state: ThreadState,
+    /// Name of the thread.
+    thread_name: Option<Vec<u8>>,
+    /// The virtual call stack.
+    stack: Vec<Frame<'mir, 'tcx, Tag, FrameData<'tcx>>>,
+    /// The join status.
+    join_status: ThreadJoinStatus,
+}
+
+impl<'mir, 'tcx> Thread<'mir, 'tcx> {
+    /// Check if the thread terminated. If yes, change the state to terminated
+    /// and return `true`.
+    fn check_terminated(&mut self) -> bool {
+        if self.state == ThreadState::Enabled {
+            if self.stack.is_empty() {
+                self.state = ThreadState::Terminated;
+                return true;
+            }
+        }
+        false
+    }
+}
+
+impl<'mir, 'tcx> std::fmt::Debug for Thread<'mir, 'tcx> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if let Some(ref name) = self.thread_name {
+            if let Ok(name_str) = std::str::from_utf8(name) {
+                write!(f, "{}", name_str)?;
+            } else {
+                write!(f, "<invalid thread name>")?;
```

Comment:
```suggestion
                write!(f, "<non-UTF-8 thread name>")?;
```
But wouldn't it be better to use `from_utf8_lossy`, and always display something?

Reply:
> But wouldn't it be better to use from_utf8_lossy, and always display something?

Thanks! I did not know about this method.